modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.cc

```diff
@@ -0,0 +1,135 @@
+// SPDX-FileCopyrightText: 2024 The OpenSn Authors <https://open-sn.github.io/opensn/>
+// SPDX-License-Identifier: MIT
+
+#include "modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/classic_richardson.h"
+#include "modules/linear_boltzmann_solvers/discrete_ordinates_solver/iterative_methods/sweep_wgs_context.h"
+#include "modules/linear_boltzmann_solvers/lbs_solver/acceleration/diffusion_mip_solver.h"
+#include "modules/linear_boltzmann_solvers/lbs_solver/lbs_solver.h"
+#include "framework/math/linear_solver/linear_solver.h"
+#include "framework/logging/log.h"
+#include "framework/utils/timer.h"
+#include "framework/runtime.h"
+#include <memory>
+#include <iomanip>
+
+namespace opensn
+{
+namespace lbs
+{
+
+ClassicRichardson::ClassicRichardson(std::shared_ptr<WGSContext> gs_context_ptr)
+  : LinearSolver("ClassicRichardson", gs_context_ptr)
+{
+}
+
+ClassicRichardson::~ClassicRichardson()
+{
+}
+
+void
+ClassicRichardson::Solve()
+{
+  auto gs_context_ptr = std::dynamic_pointer_cast<WGSContext>(context_ptr_);
+  gs_context_ptr->PreSetupCallback();
+
+  auto& groupset = gs_context_ptr->groupset_;
+  auto& lbs_solver = gs_context_ptr->lbs_solver_;
+  auto& phi_old = lbs_solver.PhiOldLocal();
+  auto& phi_new = lbs_solver.PhiNewLocal();
+  const auto scope = gs_context_ptr->lhs_src_scope_ | gs_context_ptr->rhs_src_scope_;
+  saved_q_moments_local_ = lbs_solver.QMomentsLocal();
+  psi_old_.resize(groupset.angle_agg_->GetNumDelayedAngularDOFs().first, 0.0);
+
+  double pw_phi_change_prev = 1.0;
+  bool converged = false;
+  for (int k = 0; k < groupset.max_iterations_; ++k)
+  {
+    lbs_solver.QMomentsLocal() = saved_q_moments_local_;
+    gs_context_ptr->set_source_function_(
+      groupset, lbs_solver.QMomentsLocal(), phi_old, lbs_solver.DensitiesLocal(), scope);
+    gs_context_ptr->ApplyInverseTransportOperator(scope);
+
+    // Apply WGDSA
+    if (groupset.apply_wgdsa_)
+    {
+      std::vector<double> delta_phi;
+      lbs_solver.AssembleWGDSADeltaPhiVector(groupset, phi_new - phi_old, delta_phi);
+      groupset.wgdsa_solver_->Assemble_b(delta_phi);
+      groupset.wgdsa_solver_->Solve(delta_phi);
+      lbs_solver.DisAssembleWGDSADeltaPhiVector(groupset, delta_phi, phi_new);
+    }
+
+    // Apply TGDSA
+    if (groupset.apply_tgdsa_)
+    {
+      std::vector<double> delta_phi;
+      lbs_solver.AssembleTGDSADeltaPhiVector(groupset, phi_new - phi_old, delta_phi);
+      groupset.tgdsa_solver_->Assemble_b(delta_phi);
+      groupset.tgdsa_solver_->Solve(delta_phi);
+      lbs_solver.DisAssembleTGDSADeltaPhiVector(groupset, delta_phi, phi_new);
+    }
+
+    double pw_phi_change = lbs_solver.ComputePointwiseChange(groupset);
+    double rho = (k == 0) ? 0.0 : sqrt(pw_phi_change / pw_phi_change_prev);
+    pw_phi_change_prev = pw_phi_change;
+
+    psi_new_ = groupset.angle_agg_->GetNewDelayedAngularDOFsAsSTLVector();
+    double pw_psi_change = ComputePointwisePsiChange();
```

Comment:
at first, I was confused about these 2 routines. Would you consider changing the name of the first one to reflect what PW is computed by adding `Phi`, so that it reads `ComputePointwisePhiChange` ?

Reply:
Fixed.